### PR TITLE
Fix!(parser): ensure _parse_function consumes an R_PAREN before returning

### LINKED ISSUE #2014: Bug in transpiling?
**Before you file an issue**
- Make sure you specify the "read" dialect eg. parse_one(sql, read="spark")
    - not applicable
- Check if the issue still exists on main:
    - does not

*SQLGLOT version*: `sqlglot==17.10.2`
I'm trying to raise parsing errors as suggested by the docs. Unless I'm missing something (which I might be tbh), the following code does not raise anything.
What am I doing wrong? Is the documentation up to date? 

Sorry for the inconvenience and thank you for your time.

**Fully reproducible code snippet**
Please include a fully reproducible code snippet or the input sql, dialect, and expected out

input:
```python
import sqlglot
try:
    sqlglot.transpile("SELECT foo( FROM bar")
except sqlglot.errors.ParseError as e:
    print(e.errors)
```

expected output:
```python
[{
  'description': 'Expecting )',
  'line': 1,
  'col': 16,
  'start_context': 'SELECT foo( ',
  'highlight': 'FROM',
  'end_context': ' bar',
  'into_expression': None,
}]
```

Actual output:
empty! ⚠️ 

**Official Documentation**
Please include links to official SQL documentation related to your issue.

https://github.com/tobymao/sqlglot

### PR DESCRIPTION
**Pull request by GeorgeSittas**
Fixes #2014 

cc: @awakim

### CODE REVIEW

File: sqlglot/parser.py [hunk 1]
```diff
@@ -3781,6 +3778,16 @@ def _parse_if(self) -> t.Optional[exp.Expression]:
 
         return self._parse_window(this)
 
+    def _parse_next_value_for(self) -> t.Optional[exp.Expression]:
+        if not self._match_text_seq("VALUE", "FOR"):
+            return None
```
Comment (tobymao) [inline sqlglot/parser.py:3783]: don't you need to retreat because next was consumed?

File: sqlglot/dialects/hive.py [hunk 2]
```diff
@@ -285,9 +285,7 @@ class Parser(parser.Parser):
         }
 
         def _parse_transform(self) -> exp.Transform | exp.QueryTransform:
-            args = self._parse_csv(self._parse_lambda)
-            self._match_r_paren()
-
+            args = self._parse_wrapped_csv(self._parse_lambda)
```
Comment (tobymao) [inline sqlglot/dialects/hive.py:288]: don't you need to consume a l paren here?

File: sqlglot/parser.py (hunk 1 above)
Comment (GeorgeSittas) [reply sqlglot/parser.py:3783]: both good points, will address shortly, thanks

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (GeorgeSittas) [reply sqlglot/dialects/hive.py:288]: that's handled by the `_parse_wrapped_csv` method right?

File: sqlglot/parser.py [hunk 3]
```diff
@@ -708,14 +708,10 @@ class Parser(metaclass=_Parser):
...
-        TokenType.CASE: lambda self: self._parse_case(),
-        TokenType.IF: lambda self: self._parse_if(),
-        TokenType.NEXT_VALUE_FOR: lambda self: self.expression(
-            exp.NextValueFor,
-            this=self._parse_column(),
-            order=self._match(TokenType.OVER) and self._parse_wrapped(self._parse_order),
-        ),
+        "ANY": lambda self: self.expression(exp.Any, this=self._parse_bitwise()),
+        "CASE": lambda self: self._parse_case(),
+        "IF": lambda self: self._parse_if(),
```
Comment (tobymao) [inline sqlglot/parser.py:713]: can you get rid of more of these tokens?

File: sqlglot/dialects/hive.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/dialects/hive.py:288]: ah ok

**Comment by awakim** [comment]
Wow, Thanks a lot for handling this! 

Could not hope for faster reaction!

File: tests/test_parser.py [hunk 4]
```diff
@@ -219,6 +219,9 @@ def test_parse_errors(self):
         with self.assertRaises(ParseError):
             parse_one("WITH cte AS (SELECT * FROM x)")
 
+        with self.assertRaises(ParseError):
+            parse_one("SELECT foo( FROM bar")
+
```
Comment (awakim) [inline tests/test_parser.py:224]: Nice ;)

**Comment by GeorgeSittas** [comment]
> EDIT: What is the release plan for this "fix"?

We plan to release a new version right after this is merged.

**Pull request merged**